packages/react-reconciler/src/ReactFiberContext.js

```diff
@@ -28,273 +27,308 @@ if (__DEV__) {
   warnedAboutMissingGetChildContext = {};
 }
 
-// A cursor to the current merged context object on the stack.
-let contextStackCursor: StackCursor<Object> = createCursor(emptyObject);
-// A cursor to a boolean indicating whether the context has changed.
-let didPerformWorkStackCursor: StackCursor<boolean> = createCursor(false);
-// Keep track of the previous context object that was on the stack.
-// We use this to get access to the parent context after we have already
-// pushed the next context provider, and now need to merge their contexts.
-let previousContext: Object = emptyObject;
-
-export function getUnmaskedContext(workInProgress: Fiber): Object {
-  const hasOwnContext = isContextProvider(workInProgress);
-  if (hasOwnContext) {
-    // If the fiber is a context provider itself, when we read its context
-    // we have already pushed its own child context on the stack. A context
-    // provider should not "see" its own child context. Therefore we read the
-    // previous (parent) context instead for a context provider.
-    return previousContext;
-  }
-  return contextStackCursor.current;
-}
-
-export function cacheContext(
-  workInProgress: Fiber,
-  unmaskedContext: Object,
-  maskedContext: Object,
-) {
-  const instance = workInProgress.stateNode;
-  instance.__reactInternalMemoizedUnmaskedChildContext = unmaskedContext;
-  instance.__reactInternalMemoizedMaskedChildContext = maskedContext;
-}
-
-export function getMaskedContext(
-  workInProgress: Fiber,
-  unmaskedContext: Object,
-) {
-  const type = workInProgress.type;
-  const contextTypes = type.contextTypes;
-  if (!contextTypes) {
-    return emptyObject;
+export type LegacyContext = {
+  getUnmaskedContext(workInProgress: Fiber): Object,
+  cacheContext(
+    workInProgress: Fiber,
+    unmaskedContext: Object,
+    maskedContext: Object,
+  ): void,
+  getMaskedContext(workInProgress: Fiber, unmaskedContext: Object): Object,
+  hasContextChanged(): boolean,
+  isContextConsumer(fiber: Fiber): boolean,
+  isContextProvider(fiber: Fiber): boolean,
+  popContextProvider(fiber: Fiber): void,
+  popTopLevelContextObject(fiber: Fiber): void,
+  pushTopLevelContextObject(
+    fiber: Fiber,
+    context: Object,
+    didChange: boolean,
+  ): void,
+  processChildContext(fiber: Fiber, parentContext: Object): Object,
+  pushContextProvider(workInProgress: Fiber): boolean,
+  invalidateContextProvider(workInProgress: Fiber, didChange: boolean): void,
+  findCurrentUnmaskedContext(fiber: Fiber): Object,
+};
+
+export default function(stack: Stack): LegacyContext {
```

Comment:
There's a lot of noise in this diff, but all I did here was move this into a module constructor and add a corresponding Flow type.